### ISSUE TITLE
WebRTC is not working in Dart2JS

### ISSUE THREAD

**Issue by jodinathan** [dart-lang/sdk#42796]
Tested with dart 2.8.4, full pub deps is at the end

```
import 'dart:html';

void main() async {
  var r = new RtcPeerConnection({});
  var a = await r.createOffer();

  print(a);
}
```

js_helper.dart:1343 Uncaught (in promise) TypeError: Instance of 'RTCSessionDescription': type 'UnknownJavaScriptObject' is not a subtype of type 'FutureOr<RtcSessionDescription>'
    at Object.wrapException (js_helper.dart:1343)
    at Rti._generalAsCheckImplementation [as _as] (rti.dart:915)
    at promiseToFuture_closure.call$1 (js_util.dart:157)
    at invokeClosure (js_helper.dart:2018)
    at js_helper.dart:2050

 pub deps
Dart SDK 2.8.4
guaraci_rtc 0.0.1
|-- build_resolvers 1.3.10
|   |-- analyzer...
|   |-- build...
|   |-- crypto...
|   |-- graphs...
|   |-- logging...
|   |-- path...
|   |-- pub_semver...
|   '-- package_config 1.9.3
|       |-- charcode...
|       '-- path...
|-- build_runner 1.10.0
|   |-- build_resolvers...
|   |-- pedantic...
|   |-- args 1.6.0
|   |-- async 2.4.2
|   |   '-- collection...
|   |-- build 1.3.0
|   |   |-- analyzer...
|   |   |-- async...
|   |   |-- crypto...
|   |   |-- glob...
|   |   |-- logging...
|   |   |-- meta...
|   |   |-- path...
|   |   '-- convert 2.1.1
|   |       |-- charcode...
|   |       '-- typed_data...
|   |-- build_config 0.4.2
|   |   |-- meta...
|   |   |-- path...
|   |   |-- pubspec_parse...
|   |   |-- yaml...
|   |   |-- checked_yaml 1.0.2
|   |   |   |-- json_annotation...
|   |   |   |-- source_span...
|   |   |   '-- yaml...
|   |   '-- json_annotation 3.0.1
|   |-- build_daemon 2.1.4
|   |   |-- http_multi_server...
|   |   |-- logging...
|   |   |-- path...
|   |   |-- pedantic...
|   |   |-- pool...
|   |   |-- shelf...
|   |   |-- shelf_web_socket...
|   |   |-- stream_transform...
|   |   |-- watcher...
|   |   |-- web_socket_channel...
|   |   |-- built_collection 4.3.2
|   |   |   |-- collection...
|   |   |   '-- quiver 2.1.3
|   |   |       |-- matcher...
|   |   |       '-- meta...
|   |   '-- built_value 7.1.0
|   |       |-- built_collection...
|   |       |-- collection...
|   |       |-- quiver...
|   |       '-- fixnum 0.10.11
|   |-- build_runner_core 5.2.0
|   |   |-- async...
|   |   |-- build...
|   |   |-- build_config...
|   |   |-- build_resolvers...
|   |   |-- collection...
|   |   |-- convert...
|   |   |-- crypto...
|   |   |-- glob...
|   |   |-- graphs...
|   |   |-- json_annotation...
|   |   |-- logging...
|   |   |-- meta...
|   |   |-- package_config...
|   |   |-- path...
|   |   |-- pedantic...
|   |   |-- pool...
|   |   |-- timing...
|   |   |-- watcher...
|   |   '-- yaml...
|   |-- code_builder 3.4.0
|   |   |-- built_collection...
|   |   |-- built_value...
|   |   |-- collection...
|   |   |-- meta...
|   |   '-- matcher 0.12.9
|   |       '-- stack_trace...
|   |-- collection 1.14.13
|   |-- crypto 2.1.5
|   |   |-- collection...
|   |   |-- convert...
|   |   '-- typed_data 1.2.0
|   |       '-- collection...
|   |-- dart_style 1.3.6
|   |   |-- analyzer...
|   |   |-- args...
|   |   |-- path...
|   |   '-- source_span...
|   |-- glob 1.2.0
|   |   |-- async...
|   |   |-- collection...
|   |   |-- path...
|   |   |-- pedantic...
|   |   |-- node_io 1.1.1
|   |   |   |-- path...
|   |   |   '-- node_interop 1.1.1
|   |   |       '-- js...
|   |   '-- string_scanner 1.0.5
|   |       |-- charcode...
|   |       |-- meta...
|   |       '-- source_span...
|   |-- graphs 0.2.0
|   |-- http_multi_server 2.2.0
|   |   '-- async...
|   |-- io 0.3.4
|   |   |-- meta...
|   |   |-- path...
|   |   |-- string_scanner...
|   |   '-- charcode 1.1.3
|   |-- js 0.6.2
|   |-- logging 0.11.4
|   |-- meta 1.2.2
|   |-- mime 0.9.6+3
|   |-- path 1.7.0
|   |-- pool 1.4.0
|   |   |-- async...
|   |   '-- stack_trace...
|   |-- pub_semver 1.4.4
|   |   '-- collection...
|   |-- pubspec_parse 0.1.5
|   |   |-- checked_yaml...
|   |   |-- json_annotation...
|   |   |-- pub_semver...
|   |   '-- yaml...
|   |-- shelf 0.7.7
|   |   |-- async...
|   |   |-- collection...
|   |   |-- path...
|   |   |-- stack_trace...
|   |   |-- http_parser 3.1.4
|   |   |   |-- charcode...
|   |   |   |-- collection...
|   |   |   |-- source_span...
|   |   |   |-- string_scanner...
|   |   |   '-- typed_data...
|   |   '-- stream_channel 2.0.0
|   |       '-- async...
|   |-- shelf_web_socket 0.2.3
|   |   |-- shelf...
|   |   |-- stream_channel...
|   |   '-- web_socket_channel...
|   |-- stack_trace 1.9.5
|   |   '-- path...
|   |-- stream_transform 1.2.0
|   |-- timing 0.1.1+2
|   |   '-- json_annotation...
|   |-- watcher 0.9.7+15
|   |   |-- async...
|   |   |-- path...
|   |   '-- pedantic...
|   |-- web_socket_channel 1.1.0
|   |   |-- async...
|   |   |-- crypto...
|   |   '-- stream_channel...
|   '-- yaml 2.2.1
|       |-- charcode...
|       |-- collection...
|       |-- source_span...
|       '-- string_scanner...
|-- build_web_compilers 2.11.0
|   |-- build...
|   |-- build_config...
|   |-- collection...
|   |-- glob...
|   |-- js...
|   |-- logging...
|   |-- meta...
|   |-- path...
|   |-- pool...
|   |-- stack_trace...
|   |-- analyzer 0.39.14
|   |   |-- args...
|   |   |-- charcode...
|   |   |-- collection...
|   |   |-- convert...
|   |   |-- crypto...
|   |   |-- glob...
|   |   |-- meta...
|   |   |-- package_config...
|   |   |-- path...
|   |   |-- pub_semver...
|   |   |-- source_span...
|   |   |-- watcher...
|   |   |-- yaml...
|   |   |-- _fe_analyzer_shared 6.0.0
|   |   |   '-- meta...
|   |   |-- cli_util 0.1.4
|   |   |   '-- path...
|   |   '-- html 0.14.0+3
|   |       |-- source_span...
|   |       '-- csslib 0.16.1
|   |           '-- source_span...
|   |-- archive 2.0.13
|   |   |-- args...
|   |   |-- crypto...
|   |   '-- path...
|   |-- bazel_worker 0.1.25
|   |   |-- async...
|   |   |-- pedantic...
|   |   '-- protobuf 1.0.1
|   |       '-- fixnum...
|   |-- build_modules 2.10.1
|   |   |-- analyzer...
|   |   |-- async...
|   |   |-- bazel_worker...
|   |   |-- build...
|   |   |-- build_config...
|   |   |-- collection...
|   |   |-- crypto...
|   |   |-- glob...
|   |   |-- graphs...
|   |   |-- json_annotation...
|   |   |-- logging...
|   |   |-- meta...
|   |   |-- path...
|   |   |-- pedantic...
|   |   '-- scratch_space...
|   |-- scratch_space 0.0.4+2
|   |   |-- build...
|   |   |-- crypto...
|   |   |-- path...
|   |   |-- pedantic...
|   |   '-- pool...
|   |-- source_maps 0.10.9
|   |   '-- source_span...
|   '-- source_span 1.7.0
|       |-- charcode...
|       |-- collection...
|       |-- meta...
|       |-- path...
|       '-- term_glyph 1.1.0
'-- pedantic 1.9.2
    '-- meta...

**Comment by jodinathan** [comment]
this bug just happens with -O1. 
it does not happen with -O4.
Our homolog environment uses -O1.

**Comment by mnordine** [comment]
@jodinathan Why did you close this bug? It should work regardless of the compiler flags?

**Comment by jodinathan** [comment]
@mnordine you are right, reopening

**Comment by sigmundch** [comment]
This appears to be an issue due to a missing entry on how we map native types to Dart types.

An alternative workaround besides playing with the optimization flags, is to add code that explicitly instantiates this missing type. For example, appending `new RtcSessionDescription(null)` at the end of main makes the program works.

@rakudrama @fishythefish - any thoughts? Seems like the type is only referenced as a type variable in a generic method call, could that be causing issues in our impact computation? Once you add the explicit instantiation, the main difference in the output is that we'll have an entry in `setOrUpdateInterceptorsByTag` and `setOrUpdateLeafTags` to include `RTCSessionDescription`.

**Comment by rakudrama** [comment]
The native liveness analysis does not take type parameters into account. We might consider fixing this, but it would not help here.
`promiseToFuture` is general Dart code with very indirect use of the type parameter, so it would disable native code tree-shaking to make this work. The handling of the Promise needs to be first-class and first-order to be effective at detecting a subset of live Native types.

In the past we have handled cases similar to this by `@Creates(...)` listing all the indirectly created classes as a consequence of a call.

In this case the library fails to annotate the `JS` code adequately.

```dart
  Future<RtcSessionDescription> createOffer([Map? options]) {
    var options_dict = null;
    if (options != null) {
      options_dict = convertDartToNative_Dictionary(options);
    }
    return promiseToFuture<RtcSessionDescription>(
        JS("", "#.createOffer(#)", this, options_dict));
  }
```

The `JS` call should be modified to reflect the full set of objects that could be created as a consequence of evaluating the code, something like this:
```dart
    ...
    return promiseToFuture<RtcSessionDescription>(
        JS("returns:Promise;creates:Promise|RtcSessionDescription",
            "#.createOffer(#)", this, options_dict));
  }
```

/cc @srujzs

**Comment by srujzs** [comment]
I'm a little confused: wouldn't this also be an issue in other places where we use `JS` to create a promise e.g. `MediaDevices.getUserMedia`?

**Comment by natebosch** [comment]
> the full set of objects that could be created as a consequence of evaluating the code

Will this mean that _all_ `dart:html` types which flow through a Promise are going to be considered live for any app that uses any of them? Are there code size implications?

**Comment by rakudrama** [comment]
> Will this mean that _all_ `dart:html` types which flow through a Promise are going to be considered live for any app that uses any of them? Are there code size implications?

@natebosch No, because we will 'annotate' each call that returns a Promise as creating the type(s) of the objects eventually delivered to the Promise.

**Comment by srujzs** [comment]
Resolved in #43225.
